Allow hbbft to proceed to the next round, with protocol extensions

Vagabond opened the pull request:
The HoneyBadgerBFT paper is a bit light on the details of how it moves
to the next round. Indeed, the paper assumes transactions are filtered,
reordered or retained in the buffer via some undefined mechanism.

This commit implements a mechanism to do that, as well as a mechanism
for proving f+1 nodes all agreed on this block.

When a set of transactions has been agreed on, the hbbft protocol
returns `{result, {transactions, [binary()]}}`. The application then has
an opportunity to filter/reorder/retain any transactions it chooses. It
then calls finalize_round/3 with the set of transactions it wants the
protocol to *delete* from its buffer as well as a binary representing
the block over which it wants a threshold signature.

The protocol removes the listed transactions from its buffer and then
goes about trying to obtain a signature over the block. Assuming f+1
other nodes also performed the same block construction, a threshold
signature can be obtained and the block can be shown to have majority
agreement.

vihu commented on src/hbbft.erl:
```diff
@@ -188,17 +254,50 @@ hbbft_init_test_() ->
+                                                                                         io:format("RES ~p~n", [Res]),
+                                                                                         {lists:keyreplace(J, 1, States, {J, NewS}), merge_replies(N, [{J, Res}], Replies2)}
+                                                                                 end, {NextStates2, []}, NextStates2),
+                           {NextStates3, ConvergedResults3} = do_send_outer(PenultimateReplies, PenultimateStates, sets:new()),
```
I suppose these lines can be removed now.

vihu commented on src/hbbft.erl:
```diff
@@ -93,10 +108,61 @@ handle_msg(Data = #data{round=R}, J, {dec, R, I, Share}) ->
             %% not enough shares yet
             {Data#data{dec_shares=NewShares}, ok}
     end;
+handle_msg(Data = #data{round=R}, J, {sign, R, BinShare}) ->
```
This is not part of the protocol specification as far as I can tell. Perhaps a bit of an explanation will be good. I suspect this happens after step#3 in the hbbft protocol?

![image](https://user-images.githubusercontent.com/3199183/38527339-11be0af0-3c0f-11e8-9325-0d78af463d99.png)

Vagabond commented:
I fixed a bug in a return type and added a common_test that runs different instances of hbbft protocols in their own erlang processes and verifies they can collaborate and agree on transactions (and blocks).
